compression: remove model dependency

rockwotj opened the pull request:
In reality, our models are a higher level primative than compression,
which is really a low level operation and quite independent from our
models in the kafka/redpanda world. Make the build graph reflect this
reality.

Followup from https://github.com/redpanda-data/redpanda/pull/27235


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf commented:
did we recently update our `clang_format` or something? I've also seen similar diffs between my local `clang-format` and the result from `bazel run //tools:clang_format`

rockwotj commented:
What version is your local llvm? I've been running llvm20 for a while, and we're specifically holding back clang format to reduce churn, but now that we've got something 25.2.x I think it's time to rip the bandaid off

WillemKauf reviewed (dismissed):
good stuff

nvartolomei commented on src/v/compression/compression.h:
```diff
@@ -10,11 +10,49 @@
+
+    // values below must not intersect with the value range used to encode
+    // compression codecs in kafka batch attributes.
+    producer = std::numeric_limits<std::underlying_type_t<type>>::max()
```
This line is the reason I refrained from moving this enum into compression.

It gets better when you see this too https://github.com/redpanda-data/redpanda/blob/b0e15c7db1e904ad2de7a27018cfa7dd72fa0210/src/v/compression/compression.cc#L51.

There should be 2 enums, one in model (this one) and a separate one in the compression module.

With this change now we have logical dependency arrows from everything kafka (i.e. property validation code, config subsystem) down to a compression library which is meh.

Now if I change compression code I have to rebuild model and I have to rebuild[^1] model, config, kafka, etc. when really i should have to re-build only compression and that's it. Beside this there is cognitive load. Me adding a new compression type to the base library (maybe I want to use it in iceberg) shouldn't affect the kafka protocol, property validation, config validation, metrics too[^2], etc.

[^1]: The header is pretty slim but bazel will rebuild stuff anyway if the module changes. Just tested by checking out this PR and modifying the log line in `compression.cc` and building `//src/v/config`. It rebuilds compression.cc and does one other action (didn't look closely what else it does; in release mode it may do even more work?). In dev this does not happen. The changes to compression do not force config rebuild.

[^2]: https://github.com/redpanda-data/redpanda/blob/13b9681fe558578488e76f3b56dcf14b7c586d0f/src/v/kafka/server/kafka_probe.h#L66

nvartolomei replied:
It is bad in current form too. But for this particular case, I can steelman the current mess better as it clearly shows that compression is not independent and can not change independently from the "domain model". Flipping the arrow you create that false impression and as I argued in previous comment it is a dangerous path.

rockwotj replied:
Yes agreed it should be two enums, but this is certainly a better incremental step to there.

rockwotj replied:
And in reality the compression package rarely changes

rockwotj replied:
> Now if I change compression code I have to rebuild model and I have to rebuild1 model, config, kafka, etc. when really i should have to re-build only compression and that's it.


I agree, but the prior state is no different than after this PR.

nvartolomei replied:
```
→ nv at dev in /home/nv/redpanda on git:dev
$ bazel query --notool_deps --noimplicit_deps "allpaths(//src/v/redpanda, //src/v/compression)" --output graph | wc -l
Loading: 0 packages loaded
1350
→ nv at dev in /home/nv/redpanda on git:dev
$ g sw
Switched to branch 'switch-compression'
Your branch is up to date with 'rockwotj/switch-compression'.
→ nv at dev in /home/nv/redpanda on git:switch-compression
$ bazel query --notool_deps --noimplicit_deps "allpaths(//src/v/redpanda, //src/v/compression)" --output graph | wc -l
Loading: 0 packages loaded
2208
```

And while the former can be further improved the later can not because almost everything depends on model and now on compression transitively.

dotnwat replied:
> This line is the reason I refrained from moving this enum into compression.

yep, that's what held me back initially too in an old pr

> Yes agreed it should be two enums, but this is certainly a better incremental step to there.

yeh

rockwotj replied:
Sure, but my argument is that compression hardly ever gets changed so over time there will be less recompilations in total.

rockwotj replied:
Anyways, I'll rip the bandaid off and split the enums

nvartolomei replied:
> Sure, but my argument is that compression hardly ever gets changed so over time there will be less recompilations in total.

I'm on the same page about recompilation but that wasn't my point. It was just a way of showing quantitatively that the dependency graph is wrong.

rockwotj replied:
> It was just a way of showing quantitatively that the dependency graph is wrong.

I don't think it's wrong. Having more dependencies isn't really a bad thing. The problem is that `models` is a monolith and that is a separate issue really.

I've gone ahead and split the enum because all parties agree this is better and really usage of compression is fairly isolated in the system.

rockwotj committed:
compression: split model::compression and compression::type

One is kafka protocol, the other is general compression types
independant of Redpanda semantics.

rockwotj commented:
[Force push: split the enums for compression](https://github.com/redpanda-data/redpanda/compare/0ea60349c2d71828cd1dcf17b34260b96afa5808..3e55535937791b68e7a3c27e9d40ac9d29b69f24)

nvartolomei commented on src/v/compression/compression.cc:
```diff
@@ -69,15 +64,12 @@ iobuf compressor::uncompress(const iobuf& io, type t) {
     case type::zstd:
         return internal::zstd_compressor::uncompress(io);
     default:
```
did you mean to remove the `default` case from here?

rockwotj replied:
yes for the reason you mention below

nvartolomei commented on src/v/compression/compression.cc:
```diff
@@ -99,4 +91,25 @@ ss::future<iobuf> stream_compressor::uncompress(iobuf io, type t) {
+    case type::zstd:
+        os << "zstd";
+        break;
+    default:
```
remove the default so you get compiler aid, replace breaks; with return os; then when the enum changes and do the os << unknown outside of the case?

rockwotj replied:
done

nvartolomei commented on src/v/compression/compression.h:
```diff
@@ -11,10 +11,20 @@
+    // NOTE: This is *NOT* standard snappy compression. It uses the java-snappy
+    // framing compression which is fundamentally not compatible with upstream
+    // google snappy
+    snappy,
```
nit: rename to java_snappy?

rockwotj replied:
done in force push

nvartolomei commented on src/v/kafka/protocol/kafka_batch_adapter.cc:
```diff
@@ -204,6 +204,24 @@ iobuf kafka_batch_adapter::adapt(iobuf&& kbatch) {
+        return ::compression::type::zstd;
+    default:
+        throw std::runtime_error(
+          fmt::format("Unknown compression type: {}", static_cast<int>(c)));
```
model compression still has the ostream<< operator, right? why static cast to int? let it deal with unknowns

rockwotj replied:
done

nvartolomei commented on src/v/model/batch_compression.cc:
```diff
@@ -16,6 +16,24 @@
 namespace model {
 
+namespace {
+::compression::type to_compression_type(model::compression c) {
```
do we need this duplication? 🤔 no good home for this code?

rockwotj replied:
Meh, it's only one place and TBH we could remove the `kafka_batch_adapter` because it's only used in clients that are more than a decade old.

rockwotj replied:
I'm not overly concerned about the duplication here. It feels more forced to share the code (I thought about it!)

nvartolomei reviewed (commented):
🔥

nvartolomei commented on src/v/kafka/protocol/kafka_batch_adapter.cc:
```diff
@@ -204,6 +204,24 @@ iobuf kafka_batch_adapter::adapt(iobuf&& kbatch) {
+        return ::compression::type::lz4;
+    case model::compression::zstd:
+        return ::compression::type::zstd;
+    default:
```
throw outside of the switch?

nvartolomei commented on src/v/model/batch_compression.cc:
```diff
@@ -16,6 +16,24 @@
+    case model::compression::zstd:
+        return ::compression::type::zstd;
+    default:
+        throw std::runtime_error(
```
ditto: throw outside of switch

nvartolomei commented on src/v/model/batch_compression.cc:
```diff
@@ -71,4 +89,25 @@ compress_batch(model::compression c, model::record_batch b) {
+compress_batch_sync(model::compression c, model::record_batch b) {
+    if (c == model::compression::none) {
+        vassert(
+          b.header().attrs.compression() == model::compression::none,
```
i think this can be lifted outside of the if statement? inputs here generally should always be uncompressed?

nvartolomei replied:
otherwise the `h.attrs |= c` expression also ends up with funny results

rockwotj replied:
The |= is fine: https://github.com/redpanda-data/redpanda/blob/dev/src/v/model/record.h#L388

rockwotj replied:
some users pre set the compression type

nvartolomei replied:
is it?

src/v/model/tests/record_batch_test.cc(130): fatal error: in "test_compression_methods": critical check **b.header().attrs.compression() == model::compression::snappy has failed [lz4 != snappy]**

the only valid outcomes of this code (in preference order) should be:
a) Fail the compress call because the header says that the batch is already compressed. Double compressing will make it impossible to decompress the batch later. This introduces an invariant that you must never call compress_batch with compression attrs set.
b) Assume we don't care about attrs and we trust the user that compression must be done anyway, fail if there is a miss-match between requested compression and what is in the attrs. Exception for `model::compression::none`.
c) Assume that attrs are garbage, always overwrite them.

Current behavior is none of the above. I would prefer the first one together with fixing all the call sites so that they properly update the batch header if they update the record data with uncompressed payloads. If what you say is true then it means that we pass batches around the system with incoherent attrs and they can't be trusted. So how can we write "generic" code operating on batches if whether or not we need to decompress a batch depends on some locality and temporal properties?

```diff
 src/v/model/tests/BUILD                |  1 +
 src/v/model/tests/record_batch_test.cc | 10 ++++++++++
 2 files changed, 11 insertions(+)

diff --git i/src/v/model/tests/BUILD w/src/v/model/tests/BUILD
index a07980eb75..f2a71f1154 100644
--- i/src/v/model/tests/BUILD
+++ w/src/v/model/tests/BUILD
@@ -133,6 +133,7 @@ redpanda_cc_btest(
     deps = [
         ":random",
         "//src/v/model",
+        "//src/v/model:batch_compression",
         "//src/v/test_utils:seastar_boost",
         "@boost//:test",
         "@seastar//:testing",
diff --git i/src/v/model/tests/record_batch_test.cc w/src/v/model/tests/record_batch_test.cc
index f147ec4b76..28fa76fd38 100644
--- i/src/v/model/tests/record_batch_test.cc
+++ w/src/v/model/tests/record_batch_test.cc
@@ -7,6 +7,7 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0

+#include "model/batch_compression.h"
 #include "model/record.h"
 #include "model/record_utils.h"
 #include "model/tests/random_batch.h"
@@ -119,3 +120,12 @@ SEASTAR_THREAD_TEST_CASE(test_corrupted_record_bytes) {
       b, [](model::record& r) { BOOST_CHECK(r.offset_delta() >= 0); });
     BOOST_REQUIRE_THROW(f.get(), std::out_of_range);
 }
+
+SEASTAR_THREAD_TEST_CASE(test_compression_methods) {
+    auto b = model::test::make_random_batch(model::offset(0), 10, false);
+    b.header().attrs |= model::compression::gzip;
+    b = model::compress_batch_sync(model::compression::snappy, std::move(b));
+
+    BOOST_REQUIRE_EQUAL(
+      b.header().attrs.compression(), model::compression::snappy);
+}
```

nvartolomei commented on src/v/storage/record_batch_builder.cc:
```diff
@@ -60,16 +60,12 @@ model::record_batch record_batch_builder::build() && {
-        _records = compression::compressor::compress(_records, _compression);
-    }
-
     header.reset_size_checksum_metadata(_records);
```
ditto _Don't like this line. We call it twice for the compressed case (compute crc twice). We shouldn't do that._

rockwotj replied:
I don't think we ever use a batch builder with compressed data, but will fix anyways.

BTW today in compaction, we will compute the crc twice if the batch is compressed only, and same for Wasm transforms.

nvartolomei commented on src/v/storage/record_batch_builder.cc:
```diff
@@ -78,16 +74,11 @@ ss::future<model::record_batch> record_batch_builder::build_async() && {
-          std::move(_records), _compression);
-    }
-
     header.reset_size_checksum_metadata(_records);
```
Don't like this line. We call it twice for the compressed case (compute crc twice). We shouldn't do that.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5287b0ed-0843-4907-aa66-3e43214f050c-->
<!--TEST RESULTS-->
<!--c69cadee-bd0c-4868-a11e-5c291cefde7e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70832">build#70832</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70832#0198aa3a-fcc9-4c4d-abf3-7d59c437680b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| ShardPlacementTest       | test_node_join       | {"disable_license": true}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/70832#0198aa3a-fccc-48f1-a596-dce48adf3e8c | FLAKY         | 20/21    | upstream reliability is '99.70674486803519'. current run reliability is '95.23809523809523'. drift is 4.46865 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70849">build#70849</a></summary>


| test_class               | test_method             | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePostRestartProbeTest | post_restart_probe_test | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70849#0198ac00-18f0-40f1-bb68-224ab94b4a44 | FLAKY         | 19/21    | upstream reliability is '99.55555555555556'. current run reliability is '90.47619047619048'. drift is 9.07937 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj commented:
Force push: resolve @nvartolomei's review feedback

nvartolomei reviewed (approved):
The goal of the PR has been and remaining issues with the code are unrelated to the task at hand.

This needs be addressed but can be done in follow-up if tight on time

https://github.com/redpanda-data/redpanda/pull/27244#discussion_r2278400208

rockwotj commented:
Ill send a followup to assert the batches attrs are marked as uncompressed coming in
